urllib3/sessionmanager.py

```diff
@@ -0,0 +1,87 @@
+from .util.retry import Retry
+from .request import RequestMethods, Request
+from .poolmanager import PoolManager
+from .util.context import Context
+
+
+class SessionManager(RequestMethods):
+    """
+    Allows arbitrary requests while maintaining session context across
+    those requests. Currently, that context consists of automatic
+    cookie storage and retrieval.
+
+    :param headers:
+        Headers to include with all requests, unless other headers are
+        given explcitly
+
+    :param cookie_jar:
+        A predefined :class:`CookieJar` object to use in the session;
+        if not provided, a new one will be created.
+
+    :param \**pm_kwargs:
+        Additional paramaters to be passed to the encapsulated
+        :class:`.PoolManager` or :class:`.ProxyManager` instance.
+    """
+    def __init__(self, manager, cookie_jar=None, headers=None):
+        super(SessionManager, self).__init__(headers=headers)
+        self.manager = manager
+        self.context = Context(cookie_jar)
+
+    @classmethod
+    def with_defaults(cls, manager=None, cookie_jar=None, headers=None, **kw):
+        manager_class = kw.pop('manager_class', PoolManager)
+        manager = manager or manager_class(**kw)
+        return cls(manager, cookie_jar, headers=headers)
+
+    def set_cookie_policy(self, policy):
```

Comment:
I think the big question is, "will there ever be a situation in which a SessionManager might need to choose between contexts to store information in?"

Right now, with only cookies being handled, we don't have a good grasp on what might happen next; the extraction and application is being handled wholly by the `Context`, but with n=1, it's hard to extrapolate. If we remove the convenience methods, then we're essentially making an API commitment of "There will always be one `Context` per SessionManager, which will do all processing of requests and responses."

The secondary, much less compelling, thing to think about is that by removing the wrapper methods, we have to go down to the Context level elsewhere, making for a slightly less friendly API.